llvm/include/llvm/Analysis/ScalarEvolution.h

```diff
@@ -2447,6 +2451,9 @@ class PredicatedScalarEvolution {
 
   /// The symbolic backedge taken count.
   const SCEV *SymbolicMaxBackedgeCount = nullptr;
+
+  /// The constant max trip count for the loop.
+  std::optional<unsigned> SmallConstantMaxTripCount;
```

Comment:
```suggestion
const SCEV* SmallConstantMaxTripCount = nullptr;
```

for consistency with `BackedgeCount` and `SymbolicMaxBackedgeCount` above?

Reply:
The problem with that is the max trip count is measured as an `unsigned` value. I could potentially cache the intermediate result in the function:

```
unsigned ScalarEvolution::getSmallConstantMaxTripCount(
    const Loop *L, SmallVectorImpl<const SCEVPredicate *> *Predicates) {

  const auto *MaxExitCount =
      Predicates ? getPredicatedConstantMaxBackedgeTakenCount(L, *Predicates)
                 : getConstantMaxBackedgeTakenCount(L);
  return getConstantTripCount(dyn_cast<SCEVConstant>(MaxExitCount));
}
```

i.e. `MaxExitCount`, but then you still have to call `getConstantTripCount` each time so you see less benefit from caching it. That's why I chose to use `std::optional<unsigned>` - alternatively I could use a larger type such as `uint64_t` and treat UINT64_MAX as being equivalent to `not cached yet`.

Reply:
Ah  I see! It's not worth the trouble then I think.